app.go

```diff
@@ -453,28 +464,92 @@ func ValidateApp(opts ...Option) error {
 // registered via Invoke options. See the documentation of the App struct for
 // details on the application's initialization, startup, and shutdown logic.
 func New(opts ...Option) *App {
-	logger := fxlog.DefaultLogger(os.Stderr)
-
 	app := &App{
-		log:          logger,
+		log:          new(logBuffer),
 		startTimeout: DefaultTimeout,
 		stopTimeout:  DefaultTimeout,
 	}
 
 	for _, opt := range opts {
 		opt.apply(app)
 	}
-	app.lifecycle = &lifecycleWrapper{lifecycle.New(app.log)}
 
 	app.container = dig.New(
 		dig.DeferAcyclicVerification(),
 		dig.DryRun(app.validate),
 	)
+	var connectLogger func(logger fxlog.Logger) error
+	if app.fxlogConstructor != nil {
+		// Since user supplied a custom logger, use a buffered logger to hold
+		// all messages until user supplied logger is instantiated. Then we flush
+		// those messages after fully constructing the custom logger.
+		var buffer logBuffer
+		app.log = &buffer
+		connectLogger = buffer.Connect // func(fxlog.Logger) error
+	} else {
+		// If user hasn't supplied a custom implementation of fxlog.Logger
+		// then use default one.
+		app.log = fxlog.DefaultLogger(os.Stderr)
+	}
 
 	for _, p := range app.provides {
 		app.provide(p)
 	}
 
+	if app.err != nil {
+		return app
```

Comment:
If a custom logger was provided, and this fails (because there was an error with one of the provides, like `fx.Provide(NewFoo, NewFoo)`), we won't print anything and fail silently.

We need to flush the buffer if there was a buffer, so maybe something like this:

```suggestion
	if app.err != nil {
		connectLogger(fxlog.DefaultLogger(os.Stderr))
		return app
```

Reply:
ugh, we already have a test for this case (line 513 in `app_test.go` but since it's using a spy, we didn't have a test about flushing out the actual message. i think testing this will be easier once #747 lands since i added a way to setup zap and see what exactly gets printed out. i'll address this comment after rebase.